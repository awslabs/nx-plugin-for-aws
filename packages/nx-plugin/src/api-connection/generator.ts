/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { joinPathFragments, ProjectConfiguration, Tree } from '@nx/devkit';
import { ApiConnectionGeneratorSchema } from './schema';
import trpcReactGenerator from '../trpc/react/generator';
import { hasExportDeclaration } from '../utils/ast';
import { readToml } from '../utils/toml';
import fastApiReactGenerator from '../py/fast-api/react/generator';
import { readProjectConfigurationUnqualified } from '../utils/nx';

/**
 * List of supported source and target project types for api connections
 */
const SUPPORTED_PROJECT_TYPES = [
  'ts#trpc-api',
  'py#fast-api',
  'react',
] as const;

type ProjectType = (typeof SUPPORTED_PROJECT_TYPES)[number];

type Connection = { source: ProjectType; target: ProjectType };

/**
 * Enumerates the supported project connections
 */
const SUPPORTED_CONNECTIONS = [
  { source: 'react', target: 'ts#trpc-api' },
  { source: 'react', target: 'py#fast-api' },
] satisfies Connection[];

type ConnectionKey = (typeof SUPPORTED_CONNECTIONS)[number] extends infer C
  ? C extends Connection
    ? `${C['source']} -> ${C['target']}`
    : never
  : never;

/**
 * Generators for each connection type
 */
const CONNECTION_GENERATORS = {
  'react -> ts#trpc-api': (tree, options) =>
    trpcReactGenerator(tree, {
      frontendProjectName: options.sourceProject,
      backendProjectName: options.targetProject,
    }),
  'react -> py#fast-api': (tree, options) =>
    fastApiReactGenerator(tree, {
      frontendProjectName: options.sourceProject,
      fastApiProjectName: options.targetProject,
    }),
} satisfies Record<
  ConnectionKey,
  (tree: Tree, options: ApiConnectionGeneratorSchema) => Promise<any>
>;

/**
 * Generator for a connection from a project to an API project
 */
export const apiConnectionGenerator = async (
  tree: Tree,
  options: ApiConnectionGeneratorSchema,
) => {
  const sourceType = determineProjectType(tree, options.sourceProject);
  const targetType = determineProjectType(tree, options.targetProject);

  if (!sourceType || !SUPPORTED_PROJECT_TYPES.includes(sourceType)) {
    throw new Error(
      `This generator does not support selected source project ${options.sourceProject}`,
    );
  }
  if (!targetType || !SUPPORTED_PROJECT_TYPES.includes(targetType)) {
    throw new Error(
      `This generator does not support selected target project ${options.targetProject}`,
    );
  }

  const connection = SUPPORTED_CONNECTIONS.find(
    (c) => c.source === sourceType && c.target === targetType,
  );

  if (!connection) {
    throw new Error(
      `This generator does not support a connection from ${options.sourceProject} (${sourceType}) to ${options.targetProject} (${targetType})`,
    );
  }

  return await CONNECTION_GENERATORS[
    `${connection.source} -> ${connection.target}`
  ](tree, options);
};

/**
 * Determine whether the given project is of a known project type
 */
export const determineProjectType = (
  tree: Tree,
  projectName: string,
): ProjectType | undefined => {
  const projectConfiguration = readProjectConfigurationUnqualified(
    tree,
    projectName,
  );

  // NB: if adding new checks, ensure these go from most to least specific
  // eg. react website is more specific than typescript project

  if (isTrpcApi(tree, projectConfiguration)) {
    return 'ts#trpc-api';
  }

  if (isFastApi(tree, projectConfiguration)) {
    return 'py#fast-api';
  }

  if (isReact(tree, projectConfiguration)) {
    return 'react';
  }

  return undefined;
};

const sourceRoot = (projectConfiguration: ProjectConfiguration) =>
  projectConfiguration.sourceRoot ??
  joinPathFragments(projectConfiguration.root, 'src');

const isTrpcApi = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // If the project.json says it's a trpc api, there's no need for introspection
  if ((projectConfiguration.metadata as any)?.apiType === 'trpc') {
    return true;
  }

  // Find the src/index.ts
  const indexTs = tree.read(
    joinPathFragments(sourceRoot(projectConfiguration), 'index.ts'),
    'utf-8',
  );
  if (indexTs === null) {
    return false;
  }

  // If the index file exports an AppRouter, it's a trpc api
  if (hasExportDeclaration(indexTs, 'AppRouter')) {
    return true;
  }

  // If there's a src/router.ts or src/lambdas/router.ts which exports an AppRouter, it's a trpc api
  const trpcRouter =
    tree.read(
      joinPathFragments(sourceRoot(projectConfiguration), 'router.ts'),
      'utf-8',
    ) ??
    tree.read(
      joinPathFragments(
        sourceRoot(projectConfiguration),
        'lambdas',
        'router.ts',
      ),
      'utf-8',
    );

  if (trpcRouter === null) {
    return false;
  }

  if (hasExportDeclaration(trpcRouter, 'AppRouter')) {
    return true;
  }

  return false;
};

const isReact = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // if there's a main.tsx, it's a react app
  return tree.exists(
    joinPathFragments(sourceRoot(projectConfiguration), 'main.tsx'),
  );
};

const isFastApi = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // If the project.json says it's a fast api, there's no need for introspection
  if ((projectConfiguration.metadata as any)?.apiType === 'fast-api') {
    return true;
  }

  const pyProjectPath = joinPathFragments(
    projectConfiguration.root,
    'pyproject.toml',
  );
  if (tree.exists(pyProjectPath)) {
    const pyProject = readToml(tree, pyProjectPath);
    if (((pyProject as any)?.project?.dependencies ?? []).includes('fastapi')) {
      return true;
    }
  }

  return false;
};

export default apiConnectionGenerator;
