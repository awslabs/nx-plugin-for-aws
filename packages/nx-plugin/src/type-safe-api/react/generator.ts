/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  Tree,
  GeneratorCallback,
  readProjectConfiguration,
  installPackagesTask,
  joinPathFragments,
  generateFiles,
  addDependenciesToPackageJson,
} from '@nx/devkit';
import { TypeSafeApiReactGeneratorSchema } from './schema';
import runtimeConfigGenerator from '../../cloudscape-website/runtime-config/generator';
import kebabCase from 'lodash.kebabcase';
import camelCase from 'lodash.camelcase';
import upperFirst from 'lodash.upperfirst';
import { addDefaultImportDeclarations, addJsxComponentWrapper } from '../../utils/typescript-ast';
import { toScopeAlias } from '../../utils/npm-scope';
import { withVersions } from '../../utils/versions';
import { formatFilesInSubtree } from '../../utils/format';

export async function typeSafeApiReactGenerator(
  tree: Tree,
  options: TypeSafeApiReactGeneratorSchema
): Promise<GeneratorCallback> {
  const frontendProject = readProjectConfiguration(tree, options.frontendProject);
  const hooksLibraryProject = readProjectConfiguration(tree, options.hooksLibraryProject);

  // Add runtime config provider to the app if not already present
  await runtimeConfigGenerator(tree, {
    project: options.frontendProject,
  });

  const apiName = (hooksLibraryProject.metadata as any)?.apiName;
  if (!apiName) {
    throw new Error(`${options.hooksLibraryProject} does not appear to be a Type Safe API generated hooks project`);
  }

  const apiNameKebabCase = kebabCase(apiName);
  const apiNamePascalCase = upperFirst(camelCase(apiName));

  generateFiles(tree, joinPathFragments(__dirname, 'files'), frontendProject.root, {
    apiName,
    apiNameKebabCase,
    apiNamePascalCase,
    generatedHooksLibraryPackage: toScopeAlias(options.hooksLibraryProject),
    auth: options.auth,
  });

  const mainTsxPath = joinPathFragments(
    frontendProject.sourceRoot,
    'main.tsx'
  );

  if (!tree.exists(mainTsxPath)) {
    throw new Error(
      `Could not find main.tsx in ${frontendProject.sourceRoot}`
    );
  }

  const mainTsxContents = tree.read(mainTsxPath).toString();

  const hooksProviderComponentName = `${apiNamePascalCase}HooksProvider`;

  const updatedImports = addDefaultImportDeclarations(mainTsxContents, [
    { import: hooksProviderComponentName, from: `./components/${hooksProviderComponentName}` },
  ]);

  const mainTsxUpdatedContents = addJsxComponentWrapper(updatedImports, 'App', hooksProviderComponentName);

  if (!mainTsxUpdatedContents) {
    throw new Error('Could not locate App component in main.tsx');
  }

  if (mainTsxContents !== mainTsxUpdatedContents) {
    tree.write(mainTsxPath, mainTsxUpdatedContents);
  }

  addDependenciesToPackageJson(
    tree,
    withVersions([
      '@tanstack/react-query',
      '@aws-northstar/ui',
    ]),
    {}
  );

  await formatFilesInSubtree(tree, frontendProject.root);

  return () => {
    installPackagesTask(tree);
  };
}

export default typeSafeApiReactGenerator;
