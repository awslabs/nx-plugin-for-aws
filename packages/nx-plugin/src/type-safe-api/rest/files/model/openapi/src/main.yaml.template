openapi: 3.0.0
info:
  title: "<%- namePascalCase %>"
  version: "1.0"
  description: "<%- namePascalCase %>"

paths:
  /hello:
    get:
      summary: Say hello
      operationId: SayHello
      <%_ if (handlerLanguages.length > 0) { _%>
      x-handler:
        language: "<%- handlerLanguages[0] %>"
      <%_ } _%>
      parameters:
        - name: name
          in: query
          schema:
            type: string
          required: false
          description: Name to say hello to
      responses:
        '200':
          description: Successful response
          content:
            application/json:
              schema:
                type: object
                properties:
                  message:
                    type: string
                required:
                  - message
        '400':
          description: Bad Request Error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/BadRequestErrorResponseContent'
        '403':
          description: Not Authorized Error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/NotAuthorizedErrorResponseContent'
        '404':
          description: Not Found Error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/NotFoundErrorResponseContent'
        '500':
          description: Internal Failure Error
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/InternalFailureErrorResponseContent'

components:
  schemas:
    BadRequestErrorResponseContent:
      type: object
      description: An error at the fault of the client sending invalid input
      properties:
        message:
          type: string
          description: Message with details about the error
      required:
        - message

    NotAuthorizedErrorResponseContent:
      type: object
      description: An error due to the client not being authorized to access the resource
      properties:
        message:
          type: string
          description: Message with details about the error
      required:
        - message

    NotFoundErrorResponseContent:
      type: object
      description: An error due to the client attempting to access a missing resource
      properties:
        message:
          type: string
          description: Message with details about the error
      required:
        - message

    InternalFailureErrorResponseContent:
      type: object
      description: An internal failure at the fault of the server
      properties:
        message:
          type: string
          description: Message with details about the error
      required:
        - message
