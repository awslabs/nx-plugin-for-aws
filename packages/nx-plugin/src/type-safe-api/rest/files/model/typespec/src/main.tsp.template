import "@typespec/http";
import "@typespec/openapi";
import "@typespec/openapi3";
import "./types/errors.tsp";
<%_ if (handlerLanguages.length > 0) { _%>
import "./decorators/handler.tsp";
<%_ } _%>

using Http;
using OpenAPI;

/**
 * <%- namePascalCase %>
 */
@service({ title: "<%- namePascalCase %>" })
@info({ version: "1.0" })
namespace <%- namePascalCase %>;

@get
@route("/hello")
<%_ if (handlerLanguages.length > 0) { _%>
@handler({ language: "<%- handlerLanguages[0] %>" })
<%_ } _%>
op SayHello(@query name: string):
  | {
    message: string;
  }
  | BadRequestError
  | NotAuthorizedError
  | NotFoundError
  | InternalFailureError;
