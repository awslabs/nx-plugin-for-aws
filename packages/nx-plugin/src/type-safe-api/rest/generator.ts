/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  GeneratorCallback,
  installPackagesTask,
  joinPathFragments,
  Tree,
} from '@nx/devkit';
import {
  InferredTypeSafeRestApiSchema,
  TypeSafeRestApiGeneratorSchema,
} from './schema';
import kebabCase from 'lodash.kebabcase';
import camelCase from 'lodash.camelcase';
import upperFirst from 'lodash.upperfirst';
import { join } from 'path';
import { typeSpecRestModelGenerator } from './model/typespec';
import { openApiRestModelGenerator } from './model/openapi';
import { getNpmScopePrefix, toScopeAlias } from '../../utils/npm-scope';
import { typeScriptRestRuntimeGenerator } from './runtime/typescript';
import { uniq } from '../../utils/string';
import { typeScriptRestInfrastructureGenerator } from './infrastructure/typescript';
import { typeScriptRestHandlersGenerator } from './handlers/typescript';
import { typeScriptRestHooksLibraryGenerator } from './library/hooks';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
  sharedConstructsGenerator,
} from '../../utils/shared-constructs';
import { withVersions } from '../../utils/versions';
import { writeExportDeclarations } from '../../utils/typescript-ast';

/**
 * Generates a Type Safe REST API
 */
export const typeSafeRestApiGenerator = async (
  tree: Tree,
  schema: TypeSafeRestApiGeneratorSchema
): Promise<GeneratorCallback> => {
  const nameKebabCase = kebabCase(schema.name);
  const scope = schema.scope ? `${schema.scope}/` : getNpmScopePrefix(tree);
  const fullyQualifiedApiName = `${scope}${nameKebabCase}`;
  const namePascalCase = upperFirst(camelCase(schema.name));
  const dir = join(
    schema.directory ?? '.',
    schema.subDirectory ?? nameKebabCase
  );

  const runtimeLanguages = uniq([
    schema.infrastructureLanguage,
    ...schema.handlerLanguages,
    ...schema.runtimeLanguages,
  ]);

  const options: InferredTypeSafeRestApiSchema = {
    ...schema,
    fullyQualifiedApiName,
    namePascalCase,
    nameKebabCase,
    dir,
    runtimeLanguages,
    model: {
      fullyQualifiedName: `${fullyQualifiedApiName}-model`,
      dir: join(dir, 'model'),
      outputSpecPath: join('dist', dir, 'model', 'prepared', 'api.json'),
    },
    runtime: {
      dir: join(dir, 'generated', 'runtime'),
      typescript: runtimeLanguages.includes('typescript')
        ? {
            dir: join(dir, 'generated', 'runtime', 'typescript'),
            fullyQualifiedName: `${fullyQualifiedApiName}-runtime-typescript`,
          }
        : undefined,
    },
    infrastructure: {
      dir: join(dir, 'generated', 'infrastructure'),
      typescript:
        schema.infrastructureLanguage === 'typescript'
          ? {
              dir: join(dir, 'generated', 'infrastructure', 'typescript'),
              fullyQualifiedName: `${fullyQualifiedApiName}-infrastructure-typescript`,
            }
          : undefined,
    },
    handlers: {
      dir: join(dir, 'handlers'),
      typescript: schema.handlerLanguages.includes('typescript')
        ? {
            dir: join(dir, 'handlers', 'typescript'),
            fullyQualifiedName: `${fullyQualifiedApiName}-handlers-typescript`,
            assetPath: join('dist', dir, 'handlers', 'typescript-lambda'),
          }
        : undefined,
    },
    library: {
      dir: join(dir, 'generated', 'libraries'),
      typescriptHooks: schema.libraries.includes('typescript-react-hooks')
        ? {
            dir: join(dir, 'generated', 'libraries', 'typescript-react-hooks'),
            fullyQualifiedName: `${fullyQualifiedApiName}-typescript-react-hooks`,
          }
        : undefined,
    },
  };

  // Generate the model if needed
  if (!tree.exists(options.model.dir)) {
    switch (schema.modelLanguage) {
      case 'typespec': {
        await typeSpecRestModelGenerator(tree, options);
        break;
      }
      case 'openapi': {
        await openApiRestModelGenerator(tree, options);
        break;
      }
      default: {
        throw new Error(
          `Model language ${schema.modelLanguage} is not supported!`
        );
      }
    }
  }

  if (!tree.exists(options.infrastructure.dir)) {
    switch (options.infrastructureLanguage) {
      case 'typescript': {
        await typeScriptRestInfrastructureGenerator(tree, options);
        break;
      }
      default: {
        throw new Error(
          `Infrastructure language ${options.infrastructureLanguage} is not supported!`
        );
      }
    }
  }

  for (const handlerLanguage of options.handlerLanguages) {
    switch (handlerLanguage) {
      case 'typescript': {
        if (!tree.exists(options.handlers.typescript!.dir)) {
          await typeScriptRestHandlersGenerator(tree, options);
        }
        break;
      }
      default: {
        throw new Error(
          `Handler language ${handlerLanguage} is not supported!`
        );
      }
    }
  }

  for (const runtimeLanguage of options.runtimeLanguages) {
    switch (runtimeLanguage) {
      case 'typescript': {
        if (!tree.exists(options.runtime.typescript!.dir)) {
          await typeScriptRestRuntimeGenerator(tree, options);
        }
        break;
      }
      default: {
        throw new Error(
          `Runtime language ${runtimeLanguage} is not supported!`
        );
      }
    }
  }

  for (const library of options.libraries) {
    switch (library) {
      case 'typescript-react-hooks': {
        if (!tree.exists(options.library.typescriptHooks!.dir)) {
          await typeScriptRestHooksLibraryGenerator(tree, options);
        }
        break;
      }
      default: {
        throw new Error(`Library ${library} is not supported!`);
      }
    }
  }

  // Add the shared construct for the api
  if (options.infrastructureLanguage === 'typescript') {
    await addTypeScriptSharedConstructs(tree, options);
  }

  return () => {
    installPackagesTask(tree);
  };
};

const addTypeScriptSharedConstructs = async (
  tree: Tree,
  options: InferredTypeSafeRestApiSchema
) => {
  // Call shared constructs generator for API wrapper construct
  await sharedConstructsGenerator(tree);

  const constructsPath = joinPathFragments(
    PACKAGES_DIR,
    SHARED_CONSTRUCTS_DIR,
    'src',
    options.nameKebabCase,
    'index.ts'
  );

  if (!tree.exists(constructsPath)) {
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', SHARED_CONSTRUCTS_DIR),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR),
      {
        ...options,
        generatedInfrastructurePackage: toScopeAlias(
          options.infrastructure.typescript!.fullyQualifiedName
        ),
      }
    );

    addDependenciesToPackageJson(
      tree,
      withVersions(['constructs', 'aws-cdk-lib']),
      {}
    );

    const sharedConstructsIndexTsPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'index.ts'
    );

    writeExportDeclarations(tree, sharedConstructsIndexTsPath, [
      `./${options.nameKebabCase}/index.js`,
    ]);
  }
};

export default typeSafeRestApiGenerator;
