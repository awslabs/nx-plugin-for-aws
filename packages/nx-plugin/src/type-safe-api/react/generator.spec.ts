/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from 'nx/src/devkit-testing-exports';
import { typeSafeApiReactGenerator } from './generator';

describe('type-safe-api react generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();

    // Mock frontend project configuration
    tree.write(
      'apps/frontend/project.json',
      JSON.stringify({
        name: 'frontend',
        root: 'apps/frontend',
        sourceRoot: 'apps/frontend/src',
      })
    );

    // Mock hooks library project configuration with required apiName metadata
    tree.write(
      'libs/test-api-hooks/project.json',
      JSON.stringify({
        name: 'test-api-hooks',
        root: 'libs/test-api-hooks',
        sourceRoot: 'libs/test-api-hooks/src',
        metadata: {
          apiName: 'TestApi',
        },
      })
    );

    // Mock main.tsx file
    tree.write(
      'apps/frontend/src/main.tsx',
      `
import { App } from './app';
import { BrowserRouter } from 'react-router-dom';

export function Main() {
  return <BrowserRouter><App /></BrowserRouter>;
}
`
    );
  });

  it('should generate api hooks provider files', async () => {
    await typeSafeApiReactGenerator(tree, {
      frontendProject: 'frontend',
      hooksLibraryProject: 'test-api-hooks',
      auth: 'None',
    });

    // Verify generated files
    expect(
      tree.exists('apps/frontend/src/components/TestApiHooksProvider.tsx')
    ).toBeTruthy();
    expect(
      tree.exists('apps/frontend/src/hooks/test-api/useApi.ts')
    ).toBeTruthy();

    expect(
      tree.read(
        'apps/frontend/src/components/TestApiHooksProvider.tsx',
        'utf-8'
      )
    ).toMatchSnapshot('TestApiHooksProvider.tsx');
    expect(
      tree.read(
        'apps/frontend/src/hooks/test-api/useApi.ts',
        'utf-8'
      )
    ).toMatchSnapshot('useApi.ts');
  });

  it('should modify main.tsx correctly', async () => {
    await typeSafeApiReactGenerator(tree, {
      frontendProject: 'frontend',
      hooksLibraryProject: 'test-api-hooks',
      auth: 'None',
    });

    const mainTsxContent = tree.read('apps/frontend/src/main.tsx', 'utf-8');

    // Verify imports and component wrapper were added
    expect(mainTsxContent).toContain(
      "import TestApiHooksProvider from './components/TestApiHooksProvider'"
    );
    expect(mainTsxContent).toMatchSnapshot('main.tsx');
  });

  it('should add required dependencies', async () => {
    await typeSafeApiReactGenerator(tree, {
      frontendProject: 'frontend',
      hooksLibraryProject: 'test-api-hooks',
      auth: 'None',
    });

    const packageJson = JSON.parse(tree.read('package.json', 'utf-8'));

    // Verify dependencies were added
    expect(packageJson.dependencies['@tanstack/react-query']).toBeDefined();
    expect(packageJson.dependencies['@aws-northstar/ui']).toBeDefined();
  });

  it('should throw error if hooks library project does not have apiName metadata', async () => {
    // Mock hooks library without apiName metadata
    tree.write(
      'libs/test-api-hooks/project.json',
      JSON.stringify({
        name: 'test-api-hooks',
        root: 'libs/test-api-hooks',
        sourceRoot: 'libs/test-api-hooks/src',
      })
    );

    await expect(
      typeSafeApiReactGenerator(tree, {
        frontendProject: 'frontend',
        hooksLibraryProject: 'test-api-hooks',
        auth: 'None',
      })
    ).rejects.toThrow(/does not appear to be a Type Safe API generated hooks project/);
  });

  it('should throw error if main.tsx does not exist', async () => {
    tree.delete('apps/frontend/src/main.tsx');

    await expect(
      typeSafeApiReactGenerator(tree, {
        frontendProject: 'frontend',
        hooksLibraryProject: 'test-api-hooks',
        auth: 'None',
      })
    ).rejects.toThrow();
  });

  it('should throw error if App component is not found in main.tsx', async () => {
    tree.write(
      'apps/frontend/src/main.tsx',
      `
export function Main() {
  return <BrowserRouter>No App component</BrowserRouter>;
}
`
    );

    await expect(
      typeSafeApiReactGenerator(tree, {
        frontendProject: 'frontend',
        hooksLibraryProject: 'test-api-hooks',
        auth: 'None',
      })
    ).rejects.toThrow(/Could not locate App component/);
  });

  it('should handle IAM auth option', async () => {
    await typeSafeApiReactGenerator(tree, {
      frontendProject: 'frontend',
      hooksLibraryProject: 'test-api-hooks',
      auth: 'IAM',
    });

    const hookContent = tree.read(
      'apps/frontend/src/hooks/test-api/useApi.ts',
      'utf-8'
    );

    expect(hookContent).toMatchSnapshot('useApi-IAM.ts');
  });
});
