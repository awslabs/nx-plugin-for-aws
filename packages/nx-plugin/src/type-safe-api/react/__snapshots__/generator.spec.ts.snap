// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`type-safe-api react generator > should generate api hooks provider files > TestApiHooksProvider.tsx 1`] = `
"import { DefaultApiClientProvider } from 'test-api-hooks';
import { QueryClient, MutationCache } from '@tanstack/react-query';
import React, { useMemo } from 'react';
import { errorTitle, isApiError, useApi } from '../hooks/test-api/useApi';

export const TestApiHooksProvider: React.FC<any> = ({ children }) => {
  const client = useApi();

  const queryClient = useMemo(
    () =>
      new QueryClient({
        mutationCache: new MutationCache({
          onError: (e: unknown, _variables, _context, mutation) => {
            // When an error occurs for a mutation, extract its details
            const apiError = isApiError(e) ? e : undefined;
            const header = apiError ? errorTitle(apiError.status) : 'Error';
            const content = apiError
              ? apiError.details.message
              : \`An unexpected error occurred\`;

            // Implement your own error handling logic here if needed
            // For example, you could add the error to a list of "notifications" to be displayed to the user
            console.error(header, content);
          },
        }),
      }),
    []
  );

  return client ? (
    <DefaultApiClientProvider apiClient={client} client={queryClient}>
      {children}
    </DefaultApiClientProvider>
  ) : (
    <>Hooks provider must be instantiated inside RuntimeConfigProvider</>
  );
};

export default TestApiHooksProvider;
"
`;

exports[`type-safe-api react generator > should generate api hooks provider files > useApi.ts 1`] = `
"import {
  DefaultApi,
  Configuration,
  BadRequestErrorResponseContent,
  InternalFailureErrorResponseContent,
  NotFoundErrorResponseContent,
  NotAuthorizedErrorResponseContent,
  Middleware,
  ResponseContext,
} from 'test-api-hooks';
import { useContext, useMemo } from 'react';
import { RuntimeConfigContext } from '../../components/RuntimeConfig';

export interface ApiError {
  readonly status: number;
  readonly details:
    | BadRequestErrorResponseContent
    | InternalFailureErrorResponseContent
    | NotFoundErrorResponseContent
    | NotAuthorizedErrorResponseContent;
}

export const isApiError = (e: unknown): e is ApiError =>
  !!(e && typeof e === 'object' && 'status' in e && 'details' in e);

export const errorTitle = (status: number): string => {
  switch (status) {
    case 400:
      return 'Bad Request';
    case 403:
      return 'Not Authorized';
    case 404:
      return 'Not Found';
    case 500:
      return 'Server Error';
    default:
      return 'Error';
  }
};

/**
 * Middleware for handling API errors
 */
const errorHandlingMiddleware: Middleware = {
  post: async ({ response }: ResponseContext) => {
    if (response && response.status >= 400 && response.status < 600) {
      let details;
      try {
        details = await response.json();
      } catch (e) {
        // Unable to parse response body, so continue with default error handling
        return response;
      }
      throw {
        status: response.status,
        details,
      } as ApiError;
    }
    return response;
  },
};

export const useApi = () => {
  const client = fetch;
  const runtimeContext = useContext(RuntimeConfigContext);
  const apiUrl = runtimeContext?.typeSafeApis?.['TestApi'];

  return useMemo(() => {
    return apiUrl
      ? new DefaultApi(
          new Configuration({
            basePath: apiUrl.endsWith('/') ? apiUrl.slice(0, -1) : apiUrl,
            fetchApi: client,
            middleware: [errorHandlingMiddleware],
          })
        )
      : undefined;
  }, [client, apiUrl]);
};
"
`;

exports[`type-safe-api react generator > should handle IAM auth option > useApi-IAM.ts 1`] = `
"import {
  DefaultApi,
  Configuration,
  BadRequestErrorResponseContent,
  InternalFailureErrorResponseContent,
  NotFoundErrorResponseContent,
  NotAuthorizedErrorResponseContent,
  Middleware,
  ResponseContext,
} from 'test-api-hooks';
import useSigV4Client from '@aws-northstar/ui/components/CognitoAuth/hooks/useSigv4Client';
import { useContext, useMemo } from 'react';
import { RuntimeConfigContext } from '../../components/RuntimeConfig';

export interface ApiError {
  readonly status: number;
  readonly details:
    | BadRequestErrorResponseContent
    | InternalFailureErrorResponseContent
    | NotFoundErrorResponseContent
    | NotAuthorizedErrorResponseContent;
}

export const isApiError = (e: unknown): e is ApiError =>
  !!(e && typeof e === 'object' && 'status' in e && 'details' in e);

export const errorTitle = (status: number): string => {
  switch (status) {
    case 400:
      return 'Bad Request';
    case 403:
      return 'Not Authorized';
    case 404:
      return 'Not Found';
    case 500:
      return 'Server Error';
    default:
      return 'Error';
  }
};

/**
 * Middleware for handling API errors
 */
const errorHandlingMiddleware: Middleware = {
  post: async ({ response }: ResponseContext) => {
    if (response && response.status >= 400 && response.status < 600) {
      let details;
      try {
        details = await response.json();
      } catch (e) {
        // Unable to parse response body, so continue with default error handling
        return response;
      }
      throw {
        status: response.status,
        details,
      } as ApiError;
    }
    return response;
  },
};

export const useApi = () => {
  const client = useSigV4Client();
  const runtimeContext = useContext(RuntimeConfigContext);
  const apiUrl = runtimeContext?.typeSafeApis?.['TestApi'];

  return useMemo(() => {
    return apiUrl
      ? new DefaultApi(
          new Configuration({
            basePath: apiUrl.endsWith('/') ? apiUrl.slice(0, -1) : apiUrl,
            fetchApi: client,
            middleware: [errorHandlingMiddleware],
          })
        )
      : undefined;
  }, [client, apiUrl]);
};
"
`;

exports[`type-safe-api react generator > should modify main.tsx correctly > main.tsx 1`] = `
"import TestApiHooksProvider from './components/TestApiHooksProvider';
import RuntimeConfigProvider from './components/RuntimeConfig';
import { App } from './app';
import { BrowserRouter } from 'react-router-dom';
export function Main() {
  return (
    <RuntimeConfigProvider>
      <BrowserRouter>
        <TestApiHooksProvider>
          <App />
        </TestApiHooksProvider>
      </BrowserRouter>
    </RuntimeConfigProvider>
  );
}
"
`;
